google/cloud/storage/client.h

```diff
@@ -435,6 +435,33 @@ class Client {
     raw_client_->DeleteObject(request);
   }
 
+  /**
+   * Updates the metadata in a Google Cloud Storage Object.
+   *
+   * @param bucket_name the name of the bucket that contains the object.
+   * @param object_name the name of the object.
+   * @param metadata the new metadata for the Ojbect.  The `bucket` and `name`
+   *     fields are ignored in favor of @p bucket_name and @p object_name.
```

Comment:
This seemed like odd behavior to me, so I went looking through the ObjectMetadata definition.  It looks like the only reason these are "ignored" is that we don't populate these fields in the resulting JSON when calling `ToJsonString()` (this was surprising behavior to me).  I also don't see a test below for this -- as-is, it's easily breakable if someone were to "fix a bug" and start populating the bucket and name fields in `ObjectMetadata::ToJsonString` (resulting in a 400 error from the service if each field name and its corresponding argument in this function don't match).  Regardless of how we get around this rough edge, this needs a test.

Possible workarounds:

*  It looks like some other client libraries avoid this by not letting the user pass in bucket_name and object_name as parameters, e.g. the C# client's UpdateObject method at [1].  This would require that we populate "bucket" and "name" fields in `ObjectMetadata::ToJsonString`.
*  If we want to start populating "bucket" and "name" in `ObjectMetadata::ToJsonString` (assuming they're not empty strings), we should either throw an error to the user if each pair of field/function-argument don't match, or overwrite each corresponding object metadata field with its given function argument (this would be consistent with your current docstring comment).
*  If we **don't** want to start populating "bucket" and "name" in `ObjectMetadata::ToJsonString`, I'd recommend leaving a comment in the ToJsonString method explaining why we don't populate those fields in the resulting JSON.

[1] https://googlecloudplatform.github.io/google-cloud-dotnet/docs/Google.Cloud.Storage.V1/api/Google.Cloud.Storage.V1.StorageClientImpl.html#Google_Cloud_Storage_V1_StorageClientImpl_UpdateObject_Google_Apis_Storage_v1_Data_Object_Google_Cloud_Storage_V1_UpdateObjectOptions_